v2/internal/testcommon/test_recorder_v3.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package testcommon
+
+import (
+	"bytes"
+	"io"
+	"net/http"
+	"strings"
+	"testing"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-service-operator/v2/internal/config"
+	"github.com/Azure/azure-service-operator/v2/internal/genericarmclient"
+	"github.com/go-logr/logr"
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+
+	"gopkg.in/dnaeon/go-vcr.v3/cassette"
+	"gopkg.in/dnaeon/go-vcr.v3/recorder"
+)
+
+// recorderDetailsV3 is an implementation of testRecorder using go-vcr v3.
+type recorderDetailsV3 struct {
+	cassetteName string
+	creds        azcore.TokenCredential
+	ids          AzureIDs
+	recorder     *recorder.Recorder
+	cfg          config.Values
+	log          logr.Logger
+}
+
+var (
+	nilGuid = uuid.Nil.String()
+)
+
+func newTestRecorderV3(
+	cassetteName string,
+	cfg config.Values,
+	log logr.Logger,
+) (testRecorder, error) {
+	opts := &recorder.Options{
+		CassetteName: cassetteName,
+	}
+
+	cassetteExists, err := cassetteFileExists(cassetteName)
+	if err != nil {
+		return nil, errors.Wrapf(err, "checking existence of cassette %s", cassetteName)
+	}
+
+	// Work out whether we are recording or replaying
+	if cassetteExists {
+		opts.Mode = recorder.ModeReplayOnly
+	} else {
+		opts.Mode = recorder.ModeRecordOnly
+	}
+
+	r, err := recorder.NewWithOptions(opts)
+	if err != nil {
+		return recorderDetailsV3{}, errors.Wrapf(err, "creating recorder")
+	}
+
+	var creds azcore.TokenCredential
+	var azureIDs AzureIDs
+	if r.Mode() == recorder.ModeRecordOnly {
+		// if we are recording, we need auth
+		creds, azureIDs, err = getCreds()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		// if we are replaying, we won't need auth
+		// and we use a dummy subscription ID/tenant ID
+		creds = MockTokenCredential{}
+		azureIDs.tenantID = nilGuid
+		azureIDs.subscriptionID = nilGuid
+		azureIDs.billingInvoiceID = DummyBillingId
+
+		// Force these values to be the default
+		cfg.ResourceManagerEndpoint = config.DefaultEndpoint
+		cfg.ResourceManagerAudience = config.DefaultAudience
+		cfg.AzureAuthorityHost = config.DefaultAADAuthorityHost
+	}
+
+	// check body as well as URL/Method (copied from go-vcr documentation)
+	r.SetMatcher(func(r *http.Request, i cassette.Request) bool {
+		if !cassette.DefaultMatcher(r, i) {
+			return false
+		}
+
+		// verify custom request count header (see counting_roundtripper.go)
+		if r.Header.Get(COUNT_HEADER) != i.Headers.Get(COUNT_HEADER) {
+			return false
+		}
+
+		if r.Body == nil {
+			return i.Body == ""
+		}
+
+		var b bytes.Buffer
+		if _, err := b.ReadFrom(r.Body); err != nil {
+			panic(err)
+		}
+
+		r.Body = io.NopCloser(&b)
+		return b.String() == "" || hideRecordingData(b.String()) == i.Body
+	})
+
+	r.AddHook(redactRecording(azureIDs), recorder.BeforeSaveHook)
+
+	return recorderDetailsV3{
```

Comment:
minor: Wondering if, given this method is named `new...` and `recorderDetailsV3` contains various ptr-y things, we should return `&recorderDetailsV3` here, and have the receiver be on ptr too?

Reply:
Yup.

Reply:
Done.